reduceSPI/nn.py

```diff
@@ -0,0 +1,890 @@
+            rot_mats_no_var, translation, logvar_rotmat = reparametrize_so3(
+                mu, logvar)
+        else:
+            rot_mats_no_var, translation, logvar_rotmat = reparametrize_s2s2(
+                mu, logvar)
+        rot_mats = add_logvar(rot_mats_no_var, logvar_rotmat)
+        if self.latent_space_definition == 1:
+            vec_rot = rot_mats.reshape(-1, 9)
+            z = torch.cat((vec_rot, translation), dim=1)
+            return mu, logvar, z, rot_mats
+        wig_matrix = convert_rot_mat_ten_to_wig_mat_ten(
+            rot_mats, self.latent_dim)
+        z0 = torch.matmul(wig_matrix, self.z0)
+        z1 = torch.matmul(wig_matrix, self.z1)
+        z = torch.cat((z0, z1, translation), dim=1)
+        return mu, logvar, z, rot_mats
+
+
+def reparametrize_so3(mu, logvar):
+    mu_rotmat = mu.T[:-2].T
+    mu_transl = mu.T[-2:].T
+    logvar_rotmat = logvar.T[:-2].T
+    logvar_transl = logvar.T[-2:].T
+    rot_mats_no_var = rot_mat_tensor(mu_rotmat, 3)
+    translation = reparametrize(mu_transl, logvar_transl)
+    return rot_mats_no_var, translation, logvar_rotmat
+
+
+def reparametrize_s2s2(mu, logvar):
+    mu_rotmat = mu.T[:-2].T
+    mu_transl = mu.T[-2:].T
+    logvar_rotmat = logvar.T[:-2].T
+    logvar_transl = logvar.T[-2:].T
+    translation = reparametrize(mu_transl, logvar_transl)
+    rot_mats_no_var = rotation_vectors(mu_rotmat)
+    return rot_mats_no_var, translation, logvar_rotmat
+
+
+def rotation_vectors(z):
+    z = z.reshape((-1, 6))
+    u = z.T[:3].T
+    v = z.T[3:].T
+    e1 = nn.functional.normalize(u)
+    v2 = v - (e1*v).sum(-1, keepdim=True) * e1
+    e2 = nn.functional.normalize(v2)
+    e3 = torch.cross(e1, e2)
+    matrices = torch.cat([e1, e2, e3]). reshape((-1, 3, 3))
+    return matrices
+
+
+def map_to_lie_algebra(v):
+    """Map a point in R^N to the tangent space at the identity, i.e.
+    to the Lie Algebra
+    Arg:
+        v = vector in R^N, (..., 3) in our case
+    Return:
+        R = v converted to Lie Algebra element, (3,3) in our case"""
+    # make sure this is a sample from R^3
+    assert v.shape[-1] == 3
+
+    R_x = v.new_tensor([[0., 0., 0.],
+                        [0., 0., -1.],
+                        [0., 1., 0.]])
+
+    R_y = v.new_tensor([[0., 0., 1.],
+                        [0., 0., 0.],
+                        [-1., 0., 0.]])
+
+    R_z = v.new_tensor([[0., -1., 0.],
+                        [1., 0., 0.],
+                        [0., 0., 0.]])
+
+    R = R_x * v[..., 0, None, None] + \
+        R_y * v[..., 1, None, None] + \
+        R_z * v[..., 2, None, None]
+    return R
+
+
+def rodrigues(v):
+    theta = v.norm(p=2, dim=-1, keepdim=True)
+    # normalize K
+    K = map_to_lie_algebra(v / theta)
+
+    I = torch.eye(3, device=v.device, dtype=v.dtype)
+    R = I + torch.sin(theta)[..., None]*K \
+        + (1. - torch.cos(theta))[..., None]*(K@K)
+    return R
+
+
+def add_logvar(rot_mats_no_var, logvar_rotmat):
```

Comment:
There is no sampling here, and there should be.

1) use logvar to sample a vector in R^3
2) and then rodrigues of that vector

right now you are using the value of the var to represent the vector that should be sampled.